### PR TITLE
feat: Add waitForActionAttempt option

### PR DESCRIPTION
**Pull request by razor-x**
Still need to decide on if we will provide this behavior in the client methods. However this method is a prerequisite for anything like that.

### CODE REVIEW

**Comment by razor-x** [comment]
We could introduce an option to `SeamHttp` called `awaitActionAttempts` and either codegen something like this

```ts
  async lockDoor(
    body?: LocksLockDoorBody,
  ): Promise<LocksLockDoorResponse['action_attempt']> {
    const { data } = await this.client.request<LocksLockDoorResponse>({
      url: '/locks/lock_door',
      method: 'post',
      data: body,
    })
    if (this.awaitActionAttempts) {
      return resolveActionAttempt(data.action_attempt, this)
    }
    return data.action_attempt
  }
```

Or use an axios interceptor

```ts 
  if (awaitActionAttempts) client.interceptors.response.use(actionAttemptInterceptor)

export const actionAttemptInterceptor = async <T>(response: T): Promise<T> => {
  if (isActionAttemptResponse(response)) await resolveActionAttempt()
  return response
}

```

**Comment by razor-x** [comment]
Another idea: introduce a special method:

```ts
seam.actionAttempts.resolve<T extends ActionAttempt>(params: T | Promise<T> | string): Promise<SuccessfulActionAttempt<T>>
```

I don't like this as much since it breaks parity with the API. It is kinder for tree-shaking though, 

Perhaps better is a top level method, this way the client does not need to be passed into the function

```ts
seam.resolveActionAttempt(actionAttempt)
```

This approach is not tree-shakeable though.

File: README.md [hunk 1]
```diff
@@ -145,6 +145,65 @@ const seam = SeamHttp.fromConsoleSessionToken(
...
+### Action Attempts
+
+Some asynchronous operations, e.g., unlocking a door, return an [action attempt].
+Seam tracks the progress of requested operation and updates the action attempt.
+
+To make working with action attempts more convenient for applications,
+this library provides the `resolveActionAttempt` function:
+
+```ts
+await resolveActionAttempt(seam.locks.unlockDoor({ device_id }), seam)
```
Comment (seveibar) [inline README.md:157]: Other SDKs introduce utility functions at the sdk level in various ways, many do it on the HTTP calls themselves e.g. a parameter `wait_for_action_attempt`. I think it's okay to have a utility but I don't think we want to have this call signature as the standard call signature for utility functions, what about this to start?

```suggestion
await seam.utils.resolveActionAttempt(seam.locks.unlockDoor({ device_id }))
```

I'd also recommend renaming to `waitForActionAttempt` because `resolve*` is definitely correct in promise terminology but not really how you would describe the behavior in plain english.

```suggestion
await seam.utils.waitForActionAttempt(seam.locks.unlockDoor({ device_id }))
```

Comment (razor-x) [reply README.md:157]: I actually named it `waitForActionAttempt` and then renamed it before opening this PR. 😅 We can go back. 

Adding `wait_for_action_attempt` as a param is something we can (and probably should) do: this PR is mainly to introduce the stand-alone method which is useful IMO on its own and a prerequisite for that. Introducing more options needs to be sorted out first in terms of the generator and how it's exposed, e.g., `wait_for_action_attempt` would only be valid on a subset of endpoints, and how would the non-payload params work together, e.g., how do we expose pagination?

I agree about the call signature. I saw this issue afterwards when I realized it needs a client. Opened the PR to get some ideas before refactoring.

I would like to avoid  a generic `utils` as a pattern: I always feel it's an anti pattern. AWS SDK has an example here I think we can borrow from:

```
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, GetCommand } from "@aws-sdk/lib-dynamodb";


const client = new DynamoDBClient({});
const docClient = DynamoDBDocumentClient.from(client);
```


I believe it belongs in the action attempts namespace, but since we don't want to conflate top level API endpoint methods with library methods, this could be be `SeamHttpActionAttemptsLib.fromClient(client)`, used like

```ts
await seam.actionAttempts.lib.waitForActionAttempt(seam.locks.unlockDoor({ device_id }))
```

Comment (seveibar) [reply README.md:157]: I'm not sure there's a good reason to separate HTTP-backed calls from utilities in SDK-first design. Both have network calls behind the scenes, and it's common for SDKs to rewrite the backing http call. I think the HTTP methods inform how to structure the SDK but don't need to be explicitly separated, from a consumers method they don't see the difference. So I'm recommending we remove ".lib." and agree that namespacing is good for these methods but looking at this it seems more clear that this should be a parameter and not a method, it's going to clutter the namespace a bit- can we just add it as a parameter without the utility function?

```
await seam.actionAttempts.waitForActionAttempt(seam.locks.unlockDoor({ device_id }))
```

Comment (razor-x) [reply README.md:157]: Ok, but I think we should put this in the second argument and not mixup the API request body

Comment (seveibar) [reply README.md:157]: I think it should all be in the same named argument list. Some considerations:

* Axios convention says options can be in the second argument
  * Axios options actually can change the request content, it's not a meaningful distinction and they can also be combined into the same body via some signatures
* Positions don't inherently mean anything, we're saying it can be control flow options
* Creating a conceptual understanding about what's an HTTP parameter and what's an SDK parameter doesn't help customers AFAIK, it's just an SDK w/ parameters to people who use the SDK

Comment (razor-x) [reply README.md:157]: > Axios convention says options can be in the second argument

I'm confused, is this an argument for putting it in the second option? Or are you says methods should take a second option and pass to axios?

Comment (seveibar) [reply README.md:157]: Just explaining some of my thoughts and looking for a convention around options-as-second-parameter (devil's advocate) where the first object is also an object. I think request libraries like request are the best example where this is a pattern, but I don't think Seam SDKs should adopt this pattern.

Comment (razor-x) [reply README.md:157]: Ok, I don't think I'm convinced yet but I need to think more on this. If you consider even just the case mismatch in the combined object it feels weird/like a mistake honestly. IMO the ecosystem convention here would be to separate them.

Comment (razor-x) [reply README.md:157]: Something else I remember I thought about earlier worth mentioning: some methods may not take json as the first argument (if the req body is binary). So combination will be impossible

Comment (seveibar) [reply README.md:157]: The case mismatch is a pretty good argument, I hadn't noticed that because Seam has made me case-blind, IMO that's a strong enough basis for the javascript SDK to do it as a second argument :+1:

Comment (seveibar) [reply README.md:157]: More justification: Basically we've already established we're breaking convention around data structures used by the HTTP API, I hadn't realized that (it isn't true for python where this is a named argument and also the positional approach isn't really possible)

Comment (razor-x) [reply README.md]: Ok, how does this look (see README)?

Comment (razor-x) [reply README.md]: A slight alternative option would be to do `waitForActionAttempt: bool | ResolveActionAttemptOptions`


We can add these options to the constructor as well to allow users to set a default.

**Review by codetheweb — APPROVED** [review]
API looks good :+1:

would be nice to mark generated files with

```
// .gitattributes
<glob> linguist-generated
```

as when I'm reviewing it's not immediately clear which ones were manually edited
